### PR TITLE
tweak createRandomTest params

### PR DESCRIPTION
**Pull request by cdetrio**
A bit rough, but these params work well to generate interesting traces, with many CALLs succeeding so that the call depth increases. Also the Byzantium opcodes and precompiles are weighted heavier (sampled more frequently).

### CODE REVIEW

File: test/tools/fuzzTesting/createRandomTest.cpp [hunk 1]
```diff
@@ -34,7 +34,8 @@ namespace dev { namespace test {
 bool createRandomTest()
 {
 	StateTestSuite suite;
-	dev::test::Options& options = const_cast<dev::test::Options&>(dev::test::Options::get());
+	dev::test::Options& optionsConst = const_cast<dev::test::Options&>(dev::test::Options::get());
```
Comment (winsvega) [inline test/tools/fuzzTesting/createRandomTest.cpp]: no need for cast here at all. please use dev::test::Options const&   here

File: test/tools/fuzzTesting/createRandomTest.cpp [hunk 2]
```diff
@@ -134,16 +135,20 @@ void dev::test::RandomCode::parseTestWithTypes(std::string& _test, std::map<std:
 				replace = test::RandomCode::randomUniIntHex(dev::u256("100000"), dev::u256("36028797018963967"));
 			else if (type == "[DESTADDRESS]")
 			{
-				Address address = _options.getRandomAddress(RandomCodeOptions::AddressType::PrecompiledOrStateOrCreate);
+				Address address = _options.getRandomAddress(RandomCodeOptions::AddressType::DestinationAccount);
```
Comment (winsvega) [inline test/tools/fuzzTesting/createRandomTest.cpp]: by doing this you exclude contract creation transaction and transactions directly to precompiled contracts

File: test/tools/fuzzTesting/createRandomTest.cpp [hunk 3]
```diff
@@ -134,16 +135,20 @@ void dev::test::RandomCode::parseTestWithTypes(std::string& _test, std::map<std:
...
 				if (address != ZeroAddress) //else transaction creation
 					replace = "0x" + toString(address);
 			}
-			else if (type == "[ADDRESS]")
-				replace = toString(_options.getRandomAddress(RandomCodeOptions::AddressType::StateAccount));
-			else if (type == "[0xADDRESS]")
+			else if (type == "[ADDRESS]") {
+				Address destAddress = _options.getRandomAddress(RandomCodeOptions::AddressType::StateAccount);
```
Comment (winsvega) [inline test/tools/fuzzTesting/createRandomTest.cpp]: so instead of making Options not const again. we better find a way to make test with accounts that are defined in options already set, rather then changing the options once we generate the account.

File: test/tools/fuzzTesting/createRandomTest.cpp [hunk 4]
```diff
@@ -176,8 +181,8 @@ std::vector<std::string> dev::test::RandomCode::getTypes()
 		"[0xHASH32]",			//Random hash string 0x...  32 byte length
 		"[V]",					//Random V value for transaction sig. could be invalid.
 		"[BLOCKGASLIMIT]",		//Random block gas limit with max of 2**55-1
-		"[DESTADDRESS]",		//Random destination address for transaction (could be empty string)
 		"[ADDRESS]",			//Random account address
+		"[DESTADDRESS]",		//Random destination address for transaction (could be empty string)
```
Comment (winsvega) [inline test/tools/fuzzTesting/createRandomTest.cpp]: what is the change here?

File: test/tools/fuzzTesting/fuzzHelper.cpp [hunk 5]
```diff
@@ -391,40 +399,40 @@ std::string RandomCode::fillArguments(eth::Instruction _opcode, RandomCodeOption
...
 			code += getPushCode(randUniIntGen());	//gaslimit
 			return code;
 		case eth::Instruction::SUICIDE: //(SUICIDE address)
 			code += getPushCode(toString(_options.getRandomAddress()));
 			return code;
 		case eth::Instruction::RETURN:  //(RETURN memlen1 memlen2)
 		case eth::Instruction::REVERT:  //(REVERT memlen1 memlen2)
-			code += getPushCode(randOpMemrGen());	//memlen1
-			code += getPushCode(randOpMemrGen());	//memlen1
+			code += getPushCode(randoOpSmallMemrGen());	//memlen1
```
Comment (winsvega) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: would it be better to change randOpMemrGen rather then creating a new dist?
is randOpMemrGen used anywhere now?

File: test/tools/fuzzTesting/fuzzHelper.cpp [hunk 6]
```diff
@@ -491,6 +505,10 @@ RandomCodeOptions::RandomCodeOptions() :
 	addAddress(Address("0x0000000000000000000000000000000000000006"), AddressType::Precompiled);
 	addAddress(Address("0x0000000000000000000000000000000000000007"), AddressType::Precompiled);
 	addAddress(Address("0x0000000000000000000000000000000000000008"), AddressType::Precompiled);
+	addAddress(Address("0x0000000000000000000000000000000000000005"), AddressType::ByzantiumPrecompiled);
```
Comment (winsvega) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: do we really need to distinguish precompiled from byzantium and other forks ?  
you could just initialize options with only byzantium precompiles

File: test/tools/fuzzTesting/fuzzHelper.cpp [hunk 7]
```diff
@@ -520,15 +547,42 @@ Address RandomCodeOptions::getRandomAddress(AddressType _type) const
...
 			return precompiledAddressList[(int)RandomCode::randomUniInt(0, precompiledAddressList.size())];
+		case AddressType::DestinationAccount:
+			if (RandomCode::randomPercent() < emptyAddressProbability)
+				return ZeroAddress;
+			if (test::RandomCode::randomPercent() < precompiledDestProbability)
+				return precompiledAddressList[(int)RandomCode::randomUniInt(0, precompiledAddressList.size())];
+			else {
+				if (destinationAddressList.size() == 0) {
+					return ZeroAddress;
+				} else {
```
Comment (winsvega) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: no need for  {}  here

File: test/tools/fuzzTesting/fuzzHelper.cpp [hunk 8]
```diff
@@ -520,15 +547,42 @@ Address RandomCodeOptions::getRandomAddress(AddressType _type) const
 	{
 		case AddressType::Precompiled:
 			return precompiledAddressList[(int)RandomCode::randomUniInt(0, precompiledAddressList.size())];
+		case AddressType::DestinationAccount:
+			if (RandomCode::randomPercent() < emptyAddressProbability)
+				return ZeroAddress;
```
Comment (winsvega) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: I think we should add a comment that zeroAddress means contract creation if it is a transaction

File: test/tools/fuzzTesting/createRandomTest.cpp (hunk 4 above)
Comment (cdetrio) [reply test/tools/fuzzTesting/createRandomTest.cpp]: it swaps the order that `[ADDRESS]` and `[DESTADDRESS]` are parsed in the test case template.

File: test/tools/fuzzTesting/fuzzHelper.cpp (hunk 6 above)
Comment (cdetrio) [reply test/tools/fuzzTesting/fuzzHelper.cpp]: We have to distinguish them to give Byzantium precompiles higher weights (sampling probability).

File: test/tools/fuzzTesting/createRandomTest.cpp (hunk 3 above)
Comment (cdetrio) [reply test/tools/fuzzTesting/createRandomTest.cpp]: The "constant options" restriction results in many useless/uninteresting random tests (e.g. send a tx with data to an empty address). Using this extra `destAddress` list makes test cases where the tx destination is an account with random code.

**Comment by cdetrio** [comment]
If the reason for constant options is for libfuzzer inputs https://github.com/ethereum/cpp-ethereum/issues/4515, then the `destinationAddressList` is actually fine because its deterministic (the goal of constant options for libfuzzer is deterministic test case generation, not constant options per se). Maybe we can store the `destinationAddressList` somewhere else outside of options? Then the options can stay constant.

cc @pirapira

File: test/tools/fuzzTesting/createRandomTest.cpp (hunk 3 above)
Comment (pirapira) [reply test/tools/fuzzTesting/createRandomTest.cpp]: @cdetrio which member of the `options` class do you want to change?  You can probably use the `mutable` keyword in the declaration of `options` class to achieve your goal.  Or,you can also separate the mutable thing from the options.

**Comment by pirapira** [comment]
@cdetrio constant option itself is a desired goal to keep.

File: test/tools/fuzzTesting/createRandomTest.cpp (hunk 1 above)
Comment (cdetrio) [reply test/tools/fuzzTesting/createRandomTest.cpp]: Sorry, you'll have to ELI5 (and don't know c++). The `const_cast` was already there https://github.com/ethereum/cpp-ethereum/blob/develop/test/tools/fuzzTesting/createRandomTest.cpp#L37

File: test/tools/fuzzTesting/createRandomTest.cpp (hunk 3 above)
Comment (cdetrio) [reply test/tools/fuzzTesting/createRandomTest.cpp]: @pirapira  its the `destinationAddressList` that needs to change, for the line `_options.addAddress(destAddress, RandomCodeOptions::AddressType::DestinationAccount);` to add an address to that list.

Comment (pirapira) [reply test/tools/fuzzTesting/createRandomTest.cpp]: @cdetrio you cannot change the command line options from inside the program.  You should copy the `destinationAddressList` to something mutable.

Comment (cdetrio) [reply test/tools/fuzzTesting/createRandomTest.cpp]: These address lists are not command line options, despite being members of `_options`.  I had to change `_options` to not-const in order to call [addAddress](https://github.com/ethereum/cpp-ethereum/blob/develop/test/tools/fuzzTesting/fuzzHelper.h#L61-L62) from createRandomTest.cpp. The other calls to addAddress work because they are in [fuzzHelper.cpp](https://github.com/ethereum/cpp-ethereum/blob/develop/test/tools/fuzzTesting/fuzzHelper.cpp#L480-L493), but they're not read from the command line.

Maybe all of the address lists can be moved into something mutable? Let's figure out what exactly needs to be constant, and what doesn't.

Comment (pirapira) [reply test/tools/fuzzTesting/createRandomTest.cpp]: Then the lists should move out of `Options` class.

Comment (pirapira) [reply test/tools/fuzzTesting/createRandomTest.cpp]: Hold on, it's not `Options` but `RandomCodeOptions`.  Then just say `std::vector<dev::Address> mutable destinationAddressList;`

File: test/tools/fuzzTesting/createRandomTest.cpp [hunk 9]
```diff
@@ -134,16 +135,20 @@ void dev::test::RandomCode::parseTestWithTypes(std::string& _test, std::map<std:
...
-				replace = toString(_options.getRandomAddress(RandomCodeOptions::AddressType::StateAccount));
-			else if (type == "[0xADDRESS]")
+			else if (type == "[ADDRESS]") {
+				Address destAddress = _options.getRandomAddress(RandomCodeOptions::AddressType::StateAccount);
+				_options.addAddress(destAddress, RandomCodeOptions::AddressType::DestinationAccount);
+				replace = toString(destAddress);
+			}
+			else if (type == "[0xADDRESS]") {
 				replace = "0x" + toString(_options.getRandomAddress(RandomCodeOptions::AddressType::StateAccount));
+			}
```
Comment (pirapira) [inline test/tools/fuzzTesting/createRandomTest.cpp]: Spurious `{` and `}`.  Read https://github.com/ethereum/cpp-ethereum/blob/develop/CodingStandards.txt

File: test/tools/fuzzTesting/createRandomTest.cpp (hunk 3 above)
Comment (winsvega) [reply test/tools/fuzzTesting/createRandomTest.cpp]: why not just create a template with defined accounts. then initialize options with those accounts? 
this way options would be const and there would be no calls to non existent accounts

File: test/tools/fuzzTesting/fuzzHelper.cpp (hunk 6 above)
Comment (winsvega) [reply test/tools/fuzzTesting/fuzzHelper.cpp]: you could define options with precompiled that are exist only in Byzantine. 
the goal is to make options configurable.
